src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -459,9 +401,42 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
 object TiAggregation {
   type ReturnType = PhysicalAggregation.ReturnType
 
-  def unapply(a: Any): Option[ReturnType] = a match {
+  def unapply(plan: LogicalPlan): Option[ReturnType] = plan match {
     case PhysicalAggregation(groupingExpressions, aggregateExpressions, resultExpressions, child) =>
-      Some(groupingExpressions, aggregateExpressions, resultExpressions, child)
+      // Rewrites all `Average`s into the form of `Divide(Sum / Count)` so that we can push the
+      // converted `Sum`s and `Count`s down to TiKV.
+      val averages = aggregateExpressions.collect {
+        case a @ AggregateExpression(_: Average, _, _, _) => a
+      }.distinct
+
+      val avgRewriteMap = averages.map {
+        case a @ AggregateExpression(Average(ref), _, _, _) =>
+          a.resultAttribute -> Seq(
+            a.copy(aggregateFunction = Sum(ref), resultId = newExprId),
+            a.copy(aggregateFunction = Count(ref), resultId = newExprId)
+          )
+      }.toMap
+
+      val extraAggregateExpressions = avgRewriteMap.values.reduceOption { _ ++ _ }.getOrElse(Nil)
+
+      val rewrite: PartialFunction[Expression, Expression] = avgRewriteMap.map {
+        case (ref, Seq(sum, count)) =>
+          val castedSum = Cast(sum.resultAttribute, DoubleType)
+          val castedCount = Cast(count.resultAttribute, DoubleType)
+          val division = Cast(Divide(castedSum, castedCount), ref.dataType)
+          (ref: Expression) -> Alias(division, ref.name)(exprId = ref.exprId)
+      }
+
+      val rewrittenResultExpressions = resultExpressions.map {
+        _.transform(rewrite).asInstanceOf[NamedExpression]
+      }
+
+      Some(
+        groupingExpressions,
+        (aggregateExpressions ++ extraAggregateExpressions).distinct,
+        rewrittenResultExpressions,
+        child
+      )
```

Comment:
Made `TiAggregation` always return results w/o any occurrences of `Average`. I think this is reasonable since the purpose of `TiAggregation` is to extract aggregate functions that can be pushed down to TiKV, while `Average` is not directly pushable.

Extracting the `Average` rewriting code path into a separate method also LGTM.

Reply:
Returning the original `aggregateExpressions` which may still contain an `average` expression, passing it directly to method `groupAggregateProjection`(line 388) will cause an `IllegalStateException`.

I suggest remove the `average` expression in the final `aggregateExpressions`

Reply:
My bad, nice catch!